File: src/com/jialtang/leetcode/001TwoSum.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	m := make(map[int]int)
	for i, v := range nums {
		rest := target - v
		if val, ok := m[rest]; ok {
			return []int {val, i}
		} else {
			m[v] = i
		}
	}
	return nil
}

func main() {
	var target = 5
	var nums = []int{1, 2, 3, 4}
	sum := twoSum(nums, target)
	fmt.Println(sum)
}

File: src/com/jialtang/leetcode/_001TwoSum.java
package com.jialtang.leetcode;

import java.util.HashMap;
import java.util.Map;

public class _001TwoSum {
    public static void main(String[] args) {
        int[] sums = new int[]{2, 7, 11, 15};
        int target = 9;
        int[] res = new Solution().twoSum(sums, target);

        System.out.printf("%d - %d", res[0], res[1]);
    }

    static class Solution {
        public int[] twoSum(int[] nums, int target) {
            Map<Integer, Integer> map = new HashMap();
            for (int i = 0; i < nums.length; i++) {
                int a = nums[i];
                int b = target - nums[i];
                if (map.containsKey(b)) {
                    return new int[]{map.get(b), i};
                }

                map.put(a, i);
            }

            return null;
        }
    }
}
